test/smartwallet/CustomSmartWallet.test.ts

```diff
@@ -0,0 +1,326 @@
+import {
+  CustomSmartWalletInstance,
+  TestForwarderTargetInstance,
+  TestSmartWalletInstance,
+  TestTokenInstance,
+} from '../../types/truffle-contracts'
+
+// @ts-ignore
+import { EIP712TypedData, signTypedData_v4, TypedDataUtils } from 'eth-sig-util'
+import { BN, bufferToHex, privateToAddress, toBuffer } from 'ethereumjs-util'
+import { bytes32, containsEvent, getTestingEnvironment } from '../TestUtils'
+import TypedRequestData, { ForwardRequestType, getDomainSeparatorHash } from '../../src/common/EIP712/TypedRequestData'
+import { constants } from '../../src/common/Constants'
+import { RelayRequest } from '../../src/common/EIP712/RelayRequest'
+import { expectRevert } from '@openzeppelin/test-helpers'
+
+require('source-map-support').install({ errorFormatterForce: true })
+
+const TestForwarderTarget = artifacts.require('TestForwarderTarget')
+const TestToken = artifacts.require('TestToken')
+const TestSmartWallet = artifacts.require('TestSmartWallet')
+const SuccessCustomLogic = artifacts.require('SuccessCustomLogic')
+const FailureCustomLogic = artifacts.require('FailureCustomLogic')
+const ProxyCustomLogic = artifacts.require('ProxyCustomLogic')
+const CustomSmartWallet = artifacts.require('CustomSmartWallet')
+
+async function fillTokens(token: TestTokenInstance, recipient: string, amount: string): Promise<void> {
+  await token.mint(amount, recipient);
+}
+
+async function getTokenBalance(token: TestTokenInstance, account: string): Promise<BN> {
+  return await token.balanceOf(account);
+}
+
+contract(`Custom Smart Wallet using TestToken`, ([worker, fundedAccount]) => {
+  const countParams = ForwardRequestType.length
+  const senderPrivateKey = toBuffer(bytes32(1))
+  let chainId: number
+  let senderAddress: string
+  let token: TestTokenInstance;
+  let smartWallet: CustomSmartWalletInstance
+  let domainSeparatorHash: string
+
+  const request: RelayRequest = {
+    request: {
+      relayHub: constants.ZERO_ADDRESS,
+      from: constants.ZERO_ADDRESS,
+      to: constants.ZERO_ADDRESS,
+      value: '0',
+      gas: '1000000',
+      nonce: '0',
+      data: '0x',
+      tokenContract: constants.ZERO_ADDRESS,
+      tokenAmount: '1',
+      tokenGas: '50000'
+    },
+    relayData: {
+      gasPrice: '1',
+      domainSeparator: '0x',
+      relayWorker: constants.ZERO_ADDRESS,
+      callForwarder: constants.ZERO_ADDRESS,
+      callVerifier: constants.ZERO_ADDRESS
+    }
+  }
```

Comment:
I think we could improve this, making a constructor-like function, instead of sharing the request across the context